### DISCUSSION TITLE
OAuth and Security

### DISCUSSION THREAD

**Discussion by TonyValenti** [oqtane/oqtane.framework#827] [Ideas]
To give some background, I run a small company that builds apps for legal tech.  I have noticed a common trend regarding successful legal platforms and ones that flounder: an open and secure API.

Large legal tech companies invest very heavily in their API ( [like this](https://app.clio.com/api/v4/documentation) ) because an open API allows third party developers to build extensions/integrations and that makes their platform much more "sticky".

Today, Oqtane has a public API which is both a benefit and a detriment.  Because authentication is currently controlled by only a username and password, in order to communicate with Oqtane, sensitive credentials must be shared.  Additionally, once the credentials are shared, the consuming application may now execute any action on the user's behalf.  This is a security risk.

Whether it be Clio, Rocket Matter, MyCase, CosmoLex, or a number of other legal tech companies, they all implement security/authorization using [OAuth](https://www.oauth.com/).  OAuth is really awesome because it is explicit about what data an app may or may not access and what kind of access the app has.

I believe it would be really awesome if Oqtane had OAuth built-in as it would provide a robust, secure way for developers to expose a public API to other developers.

**Comment by sbwalker** [comment]
"Support for other Auth providers" is already on the roadmap for future releases ( https://github.com/oqtane/oqtane.framework#roadmap ). Initially it was defined more specifically and mentioned integration of IdentityServer - however with the recent news that IdentityServer is adopting a commercial/GPL model it does not appear to a viable option. Since the fallout is still in progress, it would be best to wait and see the guidance Microsoft provides in this area.

**Reply by PizzaConsole** [reply]
Some stuff to read if anyone is like me and wants more info on this: https://leastprivilege.com/2020/10/01/the-future-of-identityserver/

**Reply by thabaum** [reply]
I had site DNN running with many FB G+ and Live accounts, then they changed the API, no support for fixing it, still offered in DNN but I dont think it works, everyone uses a community provided one I believe.

So I stopped using any authentication services other than what is built in.   My vote is to spend time making two factor and other security features added into octane first that surely wont change.  Anything else maybe a community supported module.

I lost all my users that connected via DNN FB Authentication. site died as no one bothered to let me know about the issue until it was pretty much too late. I do not support this at all anymore on any sites.  I stay clear of any and all dependencies.  Although convenient and people do use it.

**Reply by sbwalker** [reply]
note that the latest Visual Studio templates have switched to using the new commercial version of Identity Server - however they contain a disclaimer:

![image](https://user-images.githubusercontent.com/4840590/138154747-d6977fbd-eedb-496f-a87e-30c580066241.png)

You can review the terms of the license here: https://duendesoftware.com/license/identityserver.pdf . As mentioned previously this license is not compatible with Oqtane so we would not be able to use it.

**Reply by curlyfro** [reply]
were you considering Oqtane also hosting an Identity Server?  if not, this is ok, right?  because Oqtane just needs the OpenId client info.

**Reply by sbwalker** [reply]
@curlyfro you are correct. I was just pointing out the licensing change because a number of people in this discussion thread have mentioned IdentityServer.

**Comment by miguelhasse** [comment]
Now about integration with Azure AD B2C?
Any plans for this, or tips on how one could extend Oqtane to do so?

**Reply by sbwalker** [reply]
Support for other authentication providers (ie. OAuth) is on the roadmap but not scheduled for a specific release yet. It would need to be architected in a plug-and-play manner ( ie. the current auth system would continue to be the default, but people should be able to override it with a different implementation but preserve the functionality the framework depends upon ). I do not have any tips as I am not aware of anyone who has explored this for Oqtane yet.

**Comment by curelom** [comment]
Rather than using IdentityServer, perhaps use the new Microsoft Identity Web authentication library.

Already having current users prevents me from even considering oqtane until b2c is implemented or at least a plugin for authentication that I could use to implement it.

**Comment by ADefWebserver** [comment]
I believe that integration with Azure B2C provides the cost advantage (free for less than 50,000 users), and ease of use (works with the current Oqtane membership management code).

**Reply by markdav-is** [reply]
I'm curious about Azure B2C as well.  I see this is on the current roadmap.  Can I help out?

**Comment by sbwalker** [comment]
@curelom you are referring to:

https://developer.microsoft.com/en-us/microsoft-365/blogs/microsoft-identity-web-is-now-generally-available/

https://docs.microsoft.com/en-us/azure/active-directory/develop/microsoft-identity-web

The fact that this was quietly released in October 2020 would seem to indicate that it is intended to fill the gap created by IdentityServer?

**Reply by vzdesic** [reply]
@sbwalker do we know more details today about this topic? Do we know possible directions Oqtane could take in version 2.2.0?

I like idea of supporting AAD B2C and I will for sure use it when my application resides in the Azure, but I will also have to create my own implementation with the Identity Server 4, because my application (Composite UI, External Modules, as part of the product built according to the microservices architecture) must be able to work in completely isolated environment without any connection to the Internet (requirement for some customers). That is why I also vote for this feature to be architected in a plug-and-play manner.

I was thinking even about extending current Identity implementation in a way that it is used by Identity Server 4, but this will add a few Identity tables into Oqtane database, and this is not preferable option. Of course, I want to use Oqtane without changing default behaviour. 

I would like to know more information now about your plans here in order to plan and to do right steps.

Thanks a lot!

**Comment by cyang255** [comment]
@sbwalker I like oqtane framework, but we are California State employees, wondering if there is a separate module for validating users based on Active Directory privileges? Our next project will be using Blazor server side so I was wondering if you or anyone have any tips on incorporating oqtane with Active Directory authentication?

**Comment by sbwalker** [comment]
This feature is on the roadmap but the work has not started yet. The main focus of the next release will be ensuring that Oqtane runs properly on .NET 6. One of my colleagues @ADefWebserver  has written a very good article on integrating Blazor Server with Azure B2C: https://blazorhelpwebsite.com/ViewBlogPost/42

**Reply by markagama** [reply]
This is a great step forward, but oqtane is not practical for many users until it supports a plug and play identity management system like IdentityServer, AAD B2C, Microsoft Identity Web. The next step is really delicate, as to spend time on the wrong option could delay uptake for Oqtane. To be honest, .Net 6 is not as important as getting the identity scenario sorted out.

**Comment by ADefWebserver** [comment]
(just my opinion) I am updating my [Azure B2C articles](https://blazorhelpwebsite.com/ViewBlogPost/42) for .Net 6 and while I love Azure B2C, the configuration is not simple. Also the configuration when running in Blazor WebAssembly mode [is very complicated](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-5.0). Therefore, what @sbwalker says "current auth system would continue to be the default" is definitely the best way to go.

**Reply by vzdesic** [reply]
@ADefWebserver We all know that Azure B2C is not simple in the beginning, but many of us, interested in Oqtane, already developed some software products based on microservices architecture around Azure AD. More than anyone else, we would like to have modular web framework (micro-frontends) above our microservices architecture. We would expect from you or Team around Oqtane to start discussion and to include us, like you did it before, regarding Multiple Database Support implementation for example. Here is whole and very detailed and open discussion about that: https://github.com/oqtane/oqtane.framework/discussions/964. This is missing now, and we just see that implementation of this feature is postponed (which I agree is not proper term for open source project) release after release. If we would have opportunity to be part of that discussion then we would be able to help, so that we can all benefit from that on the end. Now, we don't see that discussion, we don't know where it is going, is it feasible at all and how this will work (as an plug-and-play fully customized option or something else). It will be great to know ASAP if there are some obstacles so that we can clearly say we will be able to rely on that in the future or not. All this will help us to make clear decisions. Now, we are stuck. There is potentially great framework for us, but still with one missing step to implement. From my perspective, this is last critical part missing. Other things may be important but are not so critical and can be done one way or another.

We need to talk openly and avoid situations where some questions are asked and never answered, like in this current thread or in this for example: https://github.com/oqtane/oqtane.framework/issues/504.

I also know that this is open source project and that you are helping with all your heart. Thanks for that!

Thanks a lot,
Vedran

**Reply by ADefWebserver** [reply]
@vzdesic I totally agree. Please understand that @sbwalker will "lead" the discussion (and make the final decision) just as he did with Multiple Database Support. Thanks!

**Reply by markagama** [reply]
@ADefWebserver I can't agree. Simply updating DNN to a new code base with a few tweaks is not positive.

Time has passed and there are new architectures available today. The monolithic server side solutions are disappearing and being replaced with lightweight front ends.

@sbwalker indicated earlier that Azure B2C would be in the pipeline after the .Net 6 code is available, and this is timely. I do hope that the discussion about a pure webassembly / SPA version of Oqtane can be entertained.

I was able to get the Microsoft Blazor Webassembly / SPA integrated with Azure AD Auth without too much drama and having done it once, I can now replicate. The key reasons were the Blazor template is designed to provide easy Azure integration.

The contributors to this project are doing a great job, and Oqtane is going to be important for Blazor uptake and vice versa. Personally, I would like to see a five person steering committee formed.

**Reply by sbwalker** [reply]
@markagama I am not sure if you are trolling this project or if you are actually genuinely trying to contribute to the discussion. I am saying this because some of your recent comments are highly inaccurate but could easily be misconstrued as fact by casual readers. Unfortunately this means that these inaccuracies need to be corrected - so I apologize in advance if my response comes across as defensive in any way. 

"Simply updating DNN to a new code base with a few tweaks is not positive."

Oqtane is NOT a migration of DNN - it was developed from the ground up on .NET Core and Blazor. It is a completely new modular web application framework that leverages modern web development patterns and techniques. 

"Time has passed and there are new architectures available today. The monolithic server side solutions are disappearing and being replaced with lightweight front ends."

Oqtane uses a completely different architecture than DNN. Oqtane is NOT a monolithic server-side solution. It is a client/server SPA application with a distinct front-end (using Blazor) and back-end (.NET Core). It's modular structure enables it to be extended with independently developed/deployed micro front-ends and services.

**Reply by markagama** [reply]
thank you for the response @sbwalker. Thank you for the defensive response and immediately calling me a troll. I take offense to this. I have noticed contradictions in your statements and differences with the latest Microsoft blazor template code and that found in Oqtane, so I'm trying to understand the differences.

If there was a FAQ it might be easier to identify answers to my questions.

I'll look forward to Oqtane 3 that supports .Net 6, I'm sure it will be a great step forward.

**Reply by sbwalker** [reply]
@markagama it seems that most of your confusion is coming from the Blazor WebAssembly application template in Visual Studio. I went ahead and tested the latest Blazor WebAssembly template in Visual Studio 2022:

![image](https://user-images.githubusercontent.com/4840590/138129454-0b817f79-44d4-4a8b-afab-8798536b0e25.png)

By default the "ASP.NET Core hosted" option is not selected. If you proceed with these defaults, the application created is extremely simplistic in its functionality. For example, the weather service in the application does not load data from a service - it simply loads raw JSON from a static file which is not a real world scenario. Another thing to keep in mind is that if you publish the application, the output needs to be deployed to a web server - which results in a wwwroot folder being created on the server containing a static index.html file. A web browser would then need to request this index.html file in order to load the app ( which would then transfer the DLLs from the server to the browser to run on WebAssembly ).  

![image](https://user-images.githubusercontent.com/4840590/138132747-794a7f7c-247d-48e3-91c4-7d1494f059e3.png)

If you select the "ASP.NET Core hosted" option, the application created is more realistic in terms of functionality. For example, it contains a Web API Controller that dynamically serves weather forecast information much like you would do in any real world application. 

![image](https://user-images.githubusercontent.com/4840590/138133860-c5493fe2-82d2-4d6b-9045-f701dc589e3f.png)

You will notice that the application contains distinct and separate Client, Server, and Shared projects. The Client project contains the front-end components. The Server contains the back-end services. And the Shared project contains the models shared with both the Client and Server. This is identical to the structure of Oqtane. And if you publish/deploy this application, the output is very similar to what I described above - although in this case there are back-end services as well as front-end components.  

Hopefully this helps clear up any confusion.

**Comment by curlyfro** [comment]
100%

given today's social web, oqtane should have had an open auth right out of the gate -- it would have helped with adoption.  it looks like oqtane's auth, roles and permissions were inherited from DNN.  it may have been done this way to quickly get the app out the door.

as long as "Support for other Auth providers" means a plug and play architecture, i can connect to my Identity Server, and there's a way i can pass my roles/permissions -- I'll be happy :)

**Reply by sbwalker** [reply]
Actually Oqtane was created BEFORE there was a clear auth story for Blazor - so the initial goal was to get the framework up and running with a security model which could serve the core needs of the framework that did not require extensive configuration or troubleshooting to make it usable in an environment ( ie. aligning with the "low ceremony" goal of the Philosophy ). However, even the current security model had its fair share of complexity due to the need to support both Blazor Server and Blazor WebAssembly - which was documented in this blog post - https://www.oqtane.org/Resources/Blog/PostId/527/exploring-authentication-in-blazor

**Comment by medbour** [comment]
Can we authenticate another spa client like mobile(flutter,maui) by adding
Jwt token to the current identity core service?
